pkg/controllers/management/oidcprovider/controller.go

```diff
@@ -0,0 +1,208 @@
+}
+
+func Register(ctx context.Context, wContext *wrangler.Context) {
+	oidcClient := wContext.Mgmt.OIDCClient()
+	controller := &oidcClientController{
+		secretClient:    wContext.Core.Secret(),
+		secretCache:     wContext.Core.Secret().Cache(),
+		oidcClient:      wContext.Mgmt.OIDCClient(),
+		oidcClientCache: wContext.Mgmt.OIDCClient().Cache(),
+		generator:       &randomstring.Generator{},
+	}
+	oidcClient.OnChange(ctx, "oidcclient-change", controller.onChange)
+	oidcClient.OnRemove(ctx, "oidcclient-remove", controller.onRemove)
+}
+
+// onChange sets a new client id in the status field, and creates a k8s with the client secret.
+func (c *oidcClientController) onChange(_ string, oidcClient *v3.OIDCClient) (*v3.OIDCClient, error) {
+	if oidcClient == nil {
+		return nil, nil
+	}
+
+	clientID := oidcClient.Status.ClientID
+
+	// generate client id
+	if oidcClient.Status.ClientID == "" {
+		var err error
+		clientID, err = c.generator.GenerateClientID()
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate clientID: %v", err)
+		}
+
+		clients, err := c.oidcClientCache.List(labels.Everything())
+		for _, client := range clients {
+			if client.Status.ClientID == clientID {
+				return nil, fmt.Errorf("client id already exists")
+			}
+		}
+		patchData := map[string]interface{}{
+			"status": map[string]string{
+				"clientID": clientID,
+			},
+		}
+
+		patchBytes, err := json.Marshal(patchData)
+		if err != nil {
+			return nil, err
+		}
+		// add client id to status
+		_, err = c.oidcClient.Patch(oidcClient.Name, types.MergePatchType, patchBytes)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	k8sSecret, err := c.secretCache.Get(secretNamespace, clientID)
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, err
+	}
+	// generate client secret and store it in a k8s secret.
+	if errors.IsNotFound(err) {
+		clientSecret, err := c.generator.GenerateClientSecret()
+		if err != nil {
+			return nil, err
+		}
+
+		k8sSecret, err = c.secretClient.Create(&v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      clientID,
+				Namespace: secretNamespace,
+			},
+			StringData: map[string]string{
+				secretKeyPrefix + "1": clientSecret,
+			},
+		})
+		if err != nil && !errors.IsAlreadyExists(err) {
+			return nil, err
+		}
+	}
+
+	// create another client secret if the cattle.io/oidc-client-secret-create annotation is present.
+	// keys are incrementing. e.g. client-secret-1, client-secret-2,...
+	if _, ok := oidcClient.Annotations[createClientSecretAnn]; ok {
+		clientSecret, err := c.generator.GenerateClientSecret()
+		if err != nil {
+			return nil, err
+		}
+		maxSecretKeyCounter := 1
+		for key, _ := range k8sSecret.Data {
+			split := strings.Split(key, "-")
+			if len(split) != 3 {
+				return nil, fmt.Errorf("invalid key found in secret")
+			}
+			num, err := strconv.Atoi(split[2])
+			if err != nil {
+				return nil, err
+			}
+			if num > maxSecretKeyCounter {
+				maxSecretKeyCounter = num
+			}
+		}
```

Comment:
Maybe move the preceding for loop into a function named `FindMaxSecretKey` or some such to make the functionality clear.

Reply:
makes sense. Added in https://github.com/rancher/rancher/pull/49558/commits/d107ca6f7c4c9e6ba0478a38cfa0808bc08d16c6